torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -256,8 +256,101 @@ class ExprFlattener : private kir::IrVisitor {
   }
 };
 
+class ValidatePlacementAfterWrites : private kir::IrVisitor {
+ public:
+  //! Validate no expr in writes found under loop
+  static void validate(
+      kir::ForLoop* loop,
+      const std::unordered_set<kir::Expr*>& writes) {
+    ValidatePlacementAfterWrites validator(writes);
+    validator.handle(loop);
+  }
+
+ private:
+  ValidatePlacementAfterWrites(const std::unordered_set<kir::Expr*>& writes)
+      : writes_(writes) {}
+
+  void handle(kir::Expr* expr) {
+    if (expr->isA<kir::ForLoop>() || expr->isA<kir::IfThenElse>()) {
+      expr->accept(this);
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          writes_.find(expr) == writes_.end(),
+          "Block sync must be placed after ",
+          kir::toString(expr));
+    }
+  }
+
+  void visit(const kir::ForLoop* fl) final {
+    for (auto expr : fl->body().exprs()) {
+      handle(expr);
+    }
+  }
+
+  void visit(const kir::IfThenElse* ite) final {
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+  }
+
+ private:
+  const std::unordered_set<kir::Expr*>& writes_;
+};
+
 class ReadAfterWriteSyncs : public kir::MutableIrVisitor {
  private:
+  //! Traverse up the loop stack from loops_it and if a halo loop is
+  //! found, place a given sync expr before the outer-most halo loop.
+  bool insertBeforeHaloLoop(
+      std::vector<kir::ForLoop*>::iterator loops_it,
+      kir::Sync* sync_expr,
+      const std::unordered_set<kir::Expr*>& writes) {
+    std::vector<kir::ForLoop*>::iterator halo_loop_it;
+    bool halo_loop_found = false;
+
+    while (true) {
+      if ((*loops_it)->iter_domain()->isThreadDim() &&
+          (*loops_it)->iter_domain()->extent() != (*loops_it)->stop()) {
+        halo_loop_found = true;
+        halo_loop_it = loops_it;
+      }
+
+      if (loops_it == for_loops_.begin()) {
+        break;
+      }
+      --loops_it;
+    }
+
+    // No halo loop found. Do not place the sync expr here. Return
+    // false to indicate nothing is done.
+    if (!halo_loop_found) {
+      return false;
+    }
+
+    auto halo_loop = *halo_loop_it;
+
+    // Make sure the writes to the smem buffer are located before the
+    // halo loop. It should be sufficient to validate no write is
+    // found under the halo loop.
+    ValidatePlacementAfterWrites::validate(halo_loop, writes);
```

Comment:
Seems this just makes sure there's no writes to the smem buffer we're trying to sync the writes of in the halo loop we need to sync before.

Reply:
Yes. Updated the comment. Hopefully it's more clear.